docs/content/ir.md

```diff
@@ -0,0 +1,435 @@
+---
+title: Intermediate Representation (IR)
+kind: misc
+weight: 1
+---
+
+# Overview
+
+OPA can compile policy queries into planned evaluation paths suitable for
+further compilation or interpretation. This document explains the structure and
+semantics of the intermediate representation (IR) used to represent these
+planned evaluation paths. Read this document if you want to write a compiler or
+interpreter for Rego.
+
+# Structure
+
+This section explains the structure of policies compiled into the IR.
+
+## Policy
+
+The root object emitted by the compiler is a `Policy` and contains the following
+top-level keys:
+
+* `static` is an object containing static data used by the compiled plans and
+  functions.
+* `plans` is an object containing entrypoints to compiled evaluation paths.
+* `funcs` is an object containing functions supporting the compiled evaluation
+  paths.
+
+## Static
+
+The `Static` object contains static data required by the plans and functions.
+The static object also contains metadata that does not affect the semantics of
+the policy. The static object contains the following top-level keys:
+
+* `strings` is an array of string constants referenced by compiled statements in
+  the plans and functions.
+* `builtin_funcs` is an array of function declarations representing built-in
+  functions required by the compiled statements.
+* `files` is used for debugging purposes only. It is an array of filenames that
+  were used during compilation.
+
+### Strings
+
+The `Strings` array is a collection of string objects referenced by compiled
+statements in the policy. Strings are referenced by their index in the
+collection. Each string object contains the following fields:
+
+* `value` is the string constant value. The string may be any valid JSON string.
+
+### Built-in Functions
+
+The `Built-in Functions` array is a collection of built-in function
+declarations. Each declaration represents a function that must be provided by
+the environment where the policy is eventually executed. Each built-in function
+contains the the following fields:
+
+* `name` is the name of the function that must be provided.
+* `decl` is the type definition of the function.
```

Comment:
@tsandall, are there by chance some tests that cover these functions? Just in case the destination language is not `golang`, the user (me) would need to ensure the logic of these built-ins re-implemented in the destination language matches the intended, standard, usage.

Maybe you have some tests hanging around after the `wasm` port?

Reply:
@dkorolev thanks for the review (and apologies for the delayed response). I've added a new "Test Suite" section at the end of the doc that links to the data-driven test suite that we use internally.